components/locale/src/extensions/transform/mod.rs

```diff
@@ -0,0 +1,156 @@
+mod fields;
+mod key;
+mod value;
+
+pub use fields::Fields;
+pub use key::Key;
+pub use value::Value;
+
+use crate::parser::{parse_language_identifier_from_iter, ParserError, ParserMode};
+
+use crate::LanguageIdentifier;
+
+use std::iter::Peekable;
+
+/// A list of [`Unicode BCP47 T Extensions`] as defined in [`Unicode Locale
+/// Identifier`] specification.
+///
+/// Transform extension carries information about source language or script of
+/// transformed content, including content that has been transliterated, transcribed,
+/// or translated, or in some other way influenced by the source (See [`RFC 6497`] for details).
+///
+/// # Examples
+///
+/// ```
+/// use icu_locale::{Locale, LanguageIdentifier};
+/// use icu_locale::extensions::transform::{Key, Value};
+///
+/// let mut loc: Locale = "de-t-en-US-h0-hybrid".parse()
+///     .expect("Parsing failed.");
+///
+/// let en_us: LanguageIdentifier = "en-US".parse()
+///     .expect("Parsing failed.");
+///
+/// assert_eq!(loc.extensions.transform.lang, Some(en_us));
+/// let key: Key = "h0".parse().expect("Parsing key failed.");
+/// let value: Value = "hybrid".parse().expect("Parsing value failed.");
+/// assert_eq!(
+///     loc.extensions.transform.fields.get(key),
+///     Some(&value)
+/// );
+/// ```
+/// [`Unicode BCP47 T Extensions`]: https://unicode.org/reports/tr35/#t_Extension
+/// [`RFC 6497`]: https://www.ietf.org/rfc/rfc6497.txt
+/// [`Unicode Locale Identifier`]: https://unicode.org/reports/tr35/#Unicode_locale_identifier
+#[derive(Clone, PartialEq, Eq, Debug, Default, Hash, PartialOrd, Ord)]
+pub struct Transform {
+    pub lang: Option<LanguageIdentifier>,
+    pub fields: Fields,
+}
+
+fn is_language_subtag(t: &[u8]) -> bool {
+    let slen = t.len();
+    (slen >= 2 && slen <= 8 || slen == 4) && !t.iter().any(|c: &u8| !c.is_ascii_alphabetic())
```

Comment:
I thought we had these boundaries defined somewhere. If we did, maybe reuse them here?

Reply:
Or maybe we refactor whole function to check if it's language subtag (code dedup).